---
title: "Finalise the localisation proportions"
author:
  - name: "Tom Smith"
    affiliation: "Cambridge Centre for Proteomics"
date: "`r format(Sys.time(), '%d %B, %Y')`"
abstract: | 
  Following the previous 2 notebooks, we now update the localisation proportions
  to generate the final object which will be used in all downstream notebooks
output:
  pdf_document:
  html_notebook: default
geometry: margin=1in
fontsize: 11pt
---

```{r}
library(biomaRt)
library(tidyverse)
library(MSnbase)
```

```{r}
ensembl <- useEnsembl(biomart="ensembl", version=102, dataset="hsapiens_gene_ensembl")

```

```{r}
proportions <- readRDS('./out/proportions_single_table.rds')

proportions_per_rep <- readRDS('./out/proportions_per_rep_single_table_flt.rds')
```


```{r}
lncRNA_mRNA_distances_min_info <- readRDS('./out/lncRNA_mRNA_distances_min_info.rds')

exclude_lncRNAs <- NULL
exclude_lncRNAs$gene <- lncRNA_mRNA_distances_min_info %>%
  filter(overlap |
           (direction=='upstream' & distance < 3E4) |
           (direction=='downstream' & distance < 1.5E4)) %>%
  pull(name_lncRNA)

gene2tx <- getBM(attributes=c('ensembl_gene_id',
                              'ensembl_transcript_id'),
                 mart = ensembl,
                 filters='ensembl_gene_id', 
                 values=exclude_lncRNAs$gene)

exclude_lncRNAs$transcript <- gene2tx$ensembl_transcript_id

exclude_lncRNAs %>% lapply(length)
```
Function to update a proportions dataframe
```{r}
update_proportions <- function(obj, exclude, recode_markers=FALSE){
  
  return_obj <- obj %>%
    filter(!id %in% exclude) %>%
    dplyr::rename(Granule=Cytosol_Light,
           sd_Granule=sd_Cytosol_Light)
  
  if(recode_markers){
    return_obj <- return_obj %>%
      mutate(markers=dplyr::recode(markers, 'Granule'='Cytosol_Light'))
  }
  
  return_obj
}

```


```{r}
proportions_per_rep %>% lapply(dim)

proportions_per_rep_final <- proportions_per_rep %>%
  names() %>%
  lapply(function(x){
    update_proportions(proportions_per_rep[[x]], exclude_lncRNAs[[x]], recode_markers=TRUE)
})

names(proportions_per_rep_final) <- names(proportions_per_rep)

proportions_per_rep_final %>% lapply(dim)
```
Update the proportions and summarise how many features there are pre/post filtering.
```{r}
proportions %>% lapply(dim)

proportions_final <- proportions %>%
  names() %>%
  lapply(function(x){
    update_proportions(proportions[[x]], exclude_lncRNAs[[x]])
})

names(proportions_final) <- names(proportions)

proportions_final %>% lapply(dim)

```
Update the quantification objects to rename Cytosol Light -> Granule
```{r}
all_datasets <- readRDS('../2_clustering/out/all_datasets_with_novel.rds')
```

```{r}
update_msnset <- function(obj){
  fData(obj)$markers[fData(obj)$markers=='Cytosol_Light'] <- 'Granule'
  obj
}

```

```{r}
all_datasets_final <- all_datasets %>% lapply(function(level){
  level %>% lapply(function(condition){
    condition %>% lapply(function(correction){
      return_obj <- update_msnset(correction) 
      return(return_obj)
    })
  })
})
```



Save final results objects
```{r}
all_datasets_final %>% saveRDS('./results/all_datasets_final.rds')
proportions_per_rep_final %>% saveRDS('./results/proportions_per_rep.rds')
proportions_final %>% saveRDS('./results/proportions.rds')
```

